includes/class-wc-product-grouped.php

```diff
@@ -10,115 +9,88 @@
 	 */
-	public function get_type() {
-		return 'grouped';
+	public function __construct( $product = 0 ) {
+		$this->data = array_merge( $this->data, $this->extra_data );
+		parent::__construct( $product );
 	}
 
+	/*
+	|--------------------------------------------------------------------------
+	| Getters
+	|--------------------------------------------------------------------------
+	|
+	| Methods for getting data from the product object.
+	*/
+
 	/**
-	 * Get the add to cart button text.
-	 *
-	 * @access public
+	 * Get internal type.
 	 * @return string
 	 */
-	public function add_to_cart_text() {
-		return apply_filters( 'woocommerce_product_add_to_cart_text', __( 'View products', 'woocommerce' ), $this );
+	public function get_type() {
+		return 'grouped';
 	}
 
 	/**
-	 * Return the products children posts.
+	 * Return the children of this product.
 	 *
-	 * @access public
 	 * @return array
 	 */
 	public function get_children() {
-		if ( ! is_array( $this->children ) || empty( $this->children ) ) {
-			$transient_name = 'wc_product_children_' . $this->id;
-			$this->children = array_filter( array_map( 'absint', (array) get_transient( $transient_name ) ) );
-
-			if ( empty( $this->children ) ) {
-
-				$args = apply_filters( 'woocommerce_grouped_children_args', array(
-					'post_parent' 	=> $this->id,
-					'post_type'		=> 'product',
-					'orderby'		=> 'menu_order',
-					'order'			=> 'ASC',
-					'fields'		=> 'ids',
-					'post_status'	=> 'publish',
-					'numberposts'	=> -1,
-				) );
-
-				$this->children = get_posts( $args );
-
-				set_transient( $transient_name, $this->children, DAY_IN_SECONDS * 30 );
-			}
-		}
-		return (array) $this->children;
+		return $this->data['children'];
 	}
 
 	/**
-	 * Returns whether or not the product has any child product.
+	 * Get the add to cart button text.
 	 *
 	 * @access public
-	 * @return bool
+	 * @return string
 	 */
-	public function has_child() {
-		return sizeof( $this->get_children() ) ? true : false;
+	public function add_to_cart_text() {
+		return apply_filters( 'woocommerce_product_add_to_cart_text', __( 'View products', 'woocommerce' ), $this );
 	}
 
 	/**
 	 * Returns whether or not the product is on sale.
 	 *
-	 * @access public
 	 * @return bool
 	 */
 	public function is_on_sale() {
-		$is_on_sale = false;
-
-		if ( $this->has_child() ) {
-
-			foreach ( $this->get_children() as $child_id ) {
-				$sale_price = get_post_meta( $child_id, '_sale_price', true );
-				if ( '' !== $sale_price && $sale_price >= 0 ) {
-					$is_on_sale = true;
-				}
-			}
-		} else {
-
-			if ( $this->sale_price && $this->sale_price == $this->price ) {
-				$is_on_sale = true;
-			}
-		}
-
-		return apply_filters( 'woocommerce_product_is_on_sale', $is_on_sale, $this );
+		global $wpdb;
+		$on_sale = 1 === $wpdb->get_var( "SELECT 1 FROM $wpdb->postmeta WHERE meta_key='_sale_price' AND meta_value>0 AND post_id IN (" . implode( ',', array_map( 'esc_sql', $this->get_children() ) ) . ");" );
```

Comment:
Can you add some spacing for this part of the query `meta_key='_sale_price' AND meta_value>0`?